doc/man/gnunetd.1
.TH GNUNETD "1" "20 Apr 2004" "GNUnet"
.SH NAME
gnunetd \- The GNUnet deamon
.SH SYNOPSIS
.B gnunetd
[\fIOPTIONS\fR]
.SH DESCRIPTION
.PP
The GNUnet deamon, required for any GNUnet operations (searches, downloads, etc). gnunetd does not require root rights.
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=FILENAME\fR
load config file (default: /etc/gnunet.conf)
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number
.TP
\fB\-d\fR, \fB\-\-debug\fR
do not detach from the console (for debugging); log messages are written to stderr instead of the logfile in -d mode.
.TP
\fB\-u \fIUSER\fR, \fB\-\-user=USER\fR
run as user USER (and if available as group USER). Note that to use this option, you will probably have to start gnunetd as
root. It is typically better to directly start gnunetd as that user instead.
.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=LOGLEVEL\fR
change the loglevel. Possible values for LOGLEVEL are NOTHING, FATAL, ERROR, FAILURE, WARNING, MESSAGE, INFO, DEBUG, CRON and EVERYTHING.
.SH NOTES
Before you can share, search or download files from GNUnet you must start the GNUnet server, gnunetd.
If you start gnunetd as root (which is not required), you can use the -u option such that GNUnet runs as a different user. If a group of the same name exists, GNUnet will also change to that group. If GNUnet can not change its UID to USER or GID to USER, a warning will be printed and gnunetd will *continue* with the rights of the user that invoked gnunetd.
.SH FILES
.TP
/etc/gnunet.conf
gnunetd configuration file (default location).  
.TP
/var/lib/GNUnet/.hostkey
Nodes GNUnet RSA private key.  Keep secret.
.TP
/var/lib/GNUnet/data/hosts/
Hostkeys.  GNUnet stores contact information and public keys of other nodes here. You may want to make this directory publically available.
.TP
/var/lib/GNUnet/data/afs/content/
Block database for anonymous file sharing. Depending on how you compiled GNUnet, a gdbm or tdb database or a directory is used.  GNUnet will store isolated blocks of files here. 
.TP
/var/lib/GNUnet/data/credit/
Trust directory.  GNUnet stores economic information about other nodes here, in particular how much useful data we have received from which remote node.
.SH "REPORTING BUGS"
Report bugs by using mantis <http://www.ovmj.org/~mantis/> or by sending electronic mail to <gnunet-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet\-gtk\fP(1), \fBgnunet\-insert\fP(1), \fBgnunet\-search\fP(1), \fBgnunet\-download\fP(1), \fBgnunet.conf\fP(5), \fBgnunetd\fP(1)


ChangeLog
Mon Feb 28 00:33:03 EST 2005
	Releasing GNUnet 0.6.6b.

Sun Feb 27 19:27:44 EST 2005
	Fixing mantis 839.

Sun Feb 27 18:44:11 EST 2005
	Fixing mantis 833/834.

Sun Feb 27 18:15:02 EST 2005
	Fixing serious problem with LE in gnunet-insert.
	Fixing mantis 828.

Thu Feb 17 01:14:30 EST 2005
	Various memory leaks for Win32 fixed.
	Added support for GNUNET(D)_CONFIG environment variable.

Wed Feb  9 13:33:38 EST 2005
        Fixed problems with the logrotate code.  Releasing GNUnet 0.6.6a.

Sun Jan 11 19:58:13 CET 2005
	Releasing GNUnet 0.6.6

Sun Jan 11 19:24:27 CET 2005
        gnunet-gtk now resumes aborted downloads automatically.

Sun Nov 14 16:49:02 EST 2004
	Debugged collections, seems to work (not that it looks pretty,
	but it basically works, the rest is UI work).  Fixed deadlock
	in gnunet-gtk logging.  Releasing GNUnet 0.6.5.

Thu Nov  4 09:04:48 EST 2004
	Added (highly experimental) support for collections.

Sat Sep 25 22:29:48 EST 2004
	Releasing GNUnet 0.6.4a.

Sun Sep 19 23:39:04 EST 2004
	Various code cleanups and minor bugfixes.

Sun Aug 29 15:11:28 IST 2004
	Releasing GNUnet 0.6.4.

Tue Aug 24 20:59:14 IST 2004
	Significant enhancements to namespaces.  A namespace can
	now be annotated with meta-data at the time that it is
	created.
	
Sat Aug 21 01:00:06 IST 2004
	Adding GNU gettext support to configure.
	
Thu Aug 19 01:26:55 IST 2004
	Migrating gnunet-gtk to GTK 2.0 (tested with 2.4).

Mon Aug 16 00:30:59 IST 2004
	Removed gcry.  Added dependency on GNU MP library (libgmp)
	for kblocks.

Tue Aug  3 17:57:43 IST 2004
	Fixed bugs in gnunet-update.  Fixed segfault in AFS startup.
	Fixed warnings in gnunet-peer-info/gnunetd.  Fixed packaging
	error (not all of util/gcry included in distribution).
	Releasing GNUnet 0.6.3a.

Mon Aug  2 20:06:07 IST 2004
	Releasing GNUnet 0.6.3.

Sun Aug  1 01:31:19 IST 2004
	Changed most places from HexName to the new, shorter EncName.
	Pushed URIs for namespaces and keyword searches into more
	widespread use.  Replaced all occurences of sprintf with
	SNPRINTF (to obtain extra bounds check).
	
Sun Jul 11 13:54:01 IST 2004
	Moved testcases for libgnunet_util to util/ directory.
	Made DB testcases in afs/module/ run for each available
	database and not just the one configured in /etc/gnunet.conf.

Fri Jun 25 21:05:59 IST 2004
	Releasing GNUnet 0.6.2b (resolves various minor bugs).

Wed May  5 17:34:39 EST 2004
	Releasing GNUnet 0.6.2a (resolves libgcrypt and packageing
	issues, improved gtk-statistics, other minor bugfixes).
	
Fri Apr 30 21:36:23 EST 2004
	Fixed synchronization problem in cron (only relevant for
	parallel downloading).  Releasing GNUnet 0.6.2.

Mon Apr 26 21:13:45 EST 2004
	Added support to gnunet-check to convert file index database
	to new design.  Removed support for pre 0.6.1b bloomfilter
	conversion (would have been too messy to support both).
	Added old-version check to gnunetd startup to make updates
	more systematic in the future (not perfect, though).

Mon Apr 26 19:27:29 EST 2004
	Recursive insertion of directories with gnunet-gtk 
	now works including the progress bar.

Sun Apr 25 18:24:55 EST 2004
	Global changes to fix bug #698.  The fix is still 
	incomplete with respect to pthread.h specific structs.
	
Sun Apr 25 15:20:29 EST 2004
	Fixed compile error in IPv6 transports.

Fri Apr 23 23:38:01 EST 2004
	Added recursive insertion of directories to gnunet-gtk.
	Progress-bar does not work yet, also libextractor is
	always used (no option to turn it off).
	
Fri Apr 23 21:49:18 EST 2004
	Removed need to specify _protocol in configuration file
	(shorter is better).

Wed Apr 21 23:32:36 EST 2004
	Added code to bound priority used by clients by twice the
	average priority of requests from other peers that are 
	routed at the moment.  That prevents clients from exposing
	themselves with excessively high priorities (and also 
	avoids giving too much credit/trust needlessly).

Tue Apr 20 22:27:19 EST 2004
	Added possibility of using a link instead of making a copy
	if gnunetd and the insert-client both run on the same machine.

Tue Apr 20 00:00:58 EST 2004
	Added network transparency for indexing files (by copying
	them to a GNUnet(d) specific directory).  This breaks
	unindexing (for files indexed with previous versions) and
	forces the user to make a copy of the file.  In the near
	future we should resolve this problem by allowing the use
	of a link.

Sun Apr 18 17:24:34 EST 2004
	Adding recursive, parallel download of directories to
	gnunet-download

Sun Apr 18 01:57:32 EST 2004
	gnunet-transport-check can now be used to check
	external connectivity (with other peers).
	
Sat Apr 17 17:46:19 EST 2004
	Reduced memory consuption for AFS insertion and
	download.
	
Sat Apr 10 21:17:46 EST 2004
	Implemented statistics-plot in gnunet-gtk based on
	gnome-system-monitor code.

Fri Apr  9 22:01:51 EST 2004
	Added support for libextractor > 0.2.7.
	
Fri Apr  9  0:29:20 EST 2004
	Completed --help conversion.

Wed Mar 31 15:40:51 EST 2004
	Releasing 0.6.1d.

Tue Mar 30 22:42:13 EST 2004
	Integrated gnunet-pseudonym-create, -delete and -list into
	gnunet-pseudonym as well as gnunet-directory-list, -emptydb
	and -print into gnunet-directory.

Tue Mar  2 10:46:01 EST 2004
	Fixing bug preventing download for indexed files <= 1k.

Sat Feb 28 03:43:34 EST 2004
	Releasing 0.6.1c.

Sat Feb 21 06:16:47 EST 2004
	Modified requestmanager to improve performance at the end of the 
	download.

Thu Feb 19 00:48:33 EST 2004
	Fixed integer underflow possibly causing slow download performance.
	
Sun Feb 15 15:17:05 EST 2004
	Fixed weak key failures with libgcrypt.

Sun Feb 15 13:14:44 EST 2004
	New routing code (untested) commited.

Sun Feb 15 11:58:10 EST 2004
	Fixing problem with 64-bit size_t on alpha (#655).

Sat Jan 31 04:32:33 EST 2004
	Releasing 0.6.1b.

Fri Jan 23 00:00:09 EST 2004
	Fixed indexed content being overwritten by migrated content
	
Thu Jan 22 19:27:33 EST 2004
	Fixed integer-overflow problems in statuscalls.

src/server/gnunetd.c
/* 
     This file is part of GNUnet.
     (C) 2001, 2002, 2004 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file server/gnunetd.c
 * @brief Daemon that must run on every GNUnet peer.
 * @author Christian Grothoff
 * @author Larry Waldo
 * @author Tzvetan Horozov
 * @author Nils Durner
 */

#include "gnunet_util.h"

#include "startup.h"
#include "handler.h"
#include "pingpong.h"
#include "heloexchange.h"
#include "knownhosts.h"
#include "tcpserver.h"
#include "core.h"
#include "traffic.h"
#include "httphelo.h"


/**
 * This flag is set if gnunetd is not (to be) detached from the
 * console.
 */
int debug_flag = NO;

/**
 * This flag is set if gnunetd was started as windows service
 */
int win_service = NO;

void gnunet_main();

#ifdef MINGW
/**
 * Windows service information
 */
SERVICE_STATUS theServiceStatus;
SERVICE_STATUS_HANDLE hService;

/**
 * This function is called from the Windows Service Control Manager
 * when a service has to shutdown
 */
void WINAPI ServiceCtrlHandler(DWORD dwOpcode) {
  if (dwOpcode == SERVICE_CONTROL_STOP)
    win_shutdown_gnunetd(SERVICE_CONTROL_STOP);
}

/**
 * Main method of the windows service
 */
void WINAPI ServiceMain(DWORD argc, LPSTR *argv) {
  memset(&theServiceStatus, sizeof(theServiceStatus), 0);
  theServiceStatus.dwServiceType = SERVICE_WIN32;
  theServiceStatus.dwControlsAccepted = SERVICE_ACCEPT_STOP;
  theServiceStatus.dwCurrentState = SERVICE_RUNNING;
  
  hService = GNRegisterServiceCtrlHandler("GNUnet", ServiceCtrlHandler);
  if (! hService)
    return;
    
  GNSetServiceStatus(hService, &theServiceStatus);
  
  gnunet_main();
  
  theServiceStatus.dwCurrentState = SERVICE_STOPPED;
  GNSetServiceStatus(hService, &theServiceStatus);
}
#endif

/**
 * The main method of gnunetd. And here is how it works:
 * <ol>
 * <li>detach from tty, initialize all coresystems
 * <li>a) start core-services
 *     b) initialize application services and download hostlist
 * <li>wait for semaphore to signal shutdown
 * <li>shutdown all services (in roughly inverse order)
 * <li>exit
 * </ol>
 */
void gnunet_main() {
  int filedes[2]; /* pipe between client and parent */
  int * sbit;
  int version;
  int firstStart;

  /* version management for GNUnet core */
  sbit = NULL;
  if (sizeof(int) == stateReadContent("GNUNET-VERSION",
				      (void**)&sbit)) {
    version = *sbit;
    FREE(sbit);
    if (ntohl(version) != 0x0630)
      errexit(_("You need to first run '%s'!\n"),
	      "gnunet-update");
    firstStart = NO;
  } else {
    FREENONNULL(sbit);  
    /* first start (or garbled version number), just write version tag */
    version = htonl(0x0630);  
    stateWriteContent("GNUNET-VERSION",
		      sizeof(int),
		      &version);
    firstStart = YES;
  }


  
  /* init 2: become deamon, initialize core subsystems */
  if (NO == debug_flag) 
    detachFromTerminal(filedes);  
  LOG(LOG_MESSAGE,
      "gnunetd starting\n");
  initHandler(); 
  initTCPServer();
  initPolicy(); 
  initTraffic();
  initKnownhosts(); 
  initConnection();   
  initPingPong();
  initCore(); 
  initTransports();
  initKeyService("gnunetd");
  initHeloExchange();  
  initHttpHelo();

  /* init 3a: start core services */
  startTransports();
  startCron();

  /* initialize signal handler (CTRL-C / SIGTERM) */
  if (NO == debug_flag) 
    detachFromTerminalComplete(filedes);  
  writePIDFile();

  /* init 3b: load application services */
  loadApplicationModules();
  if (firstStart == YES)
    downloadHostlist(); /* right away! */

  /* init 4: wait for shutdown */
  /* wait for SIGTERM, SIGTERM will set
     doShutdown to YES and send this thread
     a SIGUSR1 which will wake us up from the
     sleep */
  initSignalHandlers();
  LOG(LOG_MESSAGE,
      _("'%s' startup complete.\n"),
      "gnunetd");

  waitForSignalHandler();
  LOG(LOG_MESSAGE,
      _("'%s' is shutting down.\n"),
      "gnunetd");

  /* init 5: shutdown in inverse order */   
  disableCoreProcessing();
  stopCron();
  stopTCPServer();
  doneCore();
  deletePIDFile();
  doneHeloExchange();  
  doneHttpHelo(); 
  donePingPong();
  doneConnection();
  doneTransports();
  doneKeyService();
  doneKnownhosts();
  doneTraffic();
  doneTCPServer();
  doneHandler();
  donePolicy();
  /* init 6: goodbye */
  doneSignalHandlers();
  doneUtil();
}

/**
 * Initialize util (parse command line, options) and
 * call the main routine.
 */
int main(int argc, char * argv[]) {
  checkCompiler();
  umask(0);
  /* init 1: get options and basic services up */
  if (SYSERR == initUtil(argc, argv, &parseCommandLine))
    return 0; /* parse error, --help, etc. */

#ifdef MINGW
  if (win_service) {
    SERVICE_TABLE_ENTRY DispatchTable[] =
      {{"GNUnet", ServiceMain}, {NULL, NULL}};
    GNStartServiceCtrlDispatcher(DispatchTable);
    
    return 0;
  } else
#endif
    gnunet_main();

  return 0;
} 

/* You have reached the end of GNUnet. You can shutdown your
   computer and get a life now. */
